src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.HttpContentAsyncStream.cs

```diff
@@ -0,0 +1,412 @@
+                            PropagateError<int>(_copyTask);
+                    }
+
+                    // If there's no writer waiting for us, register our buffer/offset/count and return.
+                    if (_buffer == null)
+                    {
+                        VerboseTrace("No waiting writer. Queueing.");
+                        Update(true, buffer, offset, count, NewTcs());
+                        RegisterCancellation(cancellationToken);
+                        return _asyncOp.Task;
+                    }
+
+                    // There's writer data available to use.  Read it to satisfy this request.
+                    int bytesToCopy = Math.Min(_count, count);
+                    Debug.Assert(bytesToCopy > 0, "Expected to be copying at least 1 byte");
+                    Buffer.BlockCopy(_buffer, _offset, buffer, offset, bytesToCopy);
+                    VerboseTrace("Read " + bytesToCopy + " from writer.");
+
+                    if (bytesToCopy == _count)
+                    {
+                        // We exhausted all of the writer's data, so complete the writer
+                        // and clear out the state.
+                        TaskCompletionSource<int> tcs = _asyncOp;
+                        Update(false, null, 0, 0, null);
+                        _cancellationRegistration.Dispose();
+                        bool writerCompleted = tcs.TrySetResult(default(int)); // value doesn't matter, as the Task is a writer
+                        Debug.Assert(writerCompleted, "No one else should have completed the writer");
+                    }
+                    else
+                    {
+                        // The writer still has more data to provide to a future read.
+                        // Update based on this read.
+                        _offset += bytesToCopy;
+                        _count -= bytesToCopy;
+                    }
+
+                    // Return the number of bytes read.
+                    return Task.FromResult(bytesToCopy);
+                }
+            }
+
+            public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            {
+                // If no data was provided, we're done.
+                if (count == 0)
+                {
+                    return Task.CompletedTask;
+                }
+
+                // If cancellation was requested, bail.
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return Task.FromCanceled<int>(cancellationToken);
+                }
+
+                lock (_syncObj)
+                {
+                    // If we've been disposed, throw an exception so as to end the CopyToAsync operation.
+                    if (_disposed)
+                    {
+                        VerboseTrace("WriteAsync when already disposed");
+                        throw CreateHttpRequestException();
```

Comment:
Nit: Shall we throw IOException here so that CreateHttpRequestException usage is localized?